### ISSUE TITLE
Build on OS X is broken

### ISSUE THREAD

**Issue by tymat** [ethereum/webthree-umbrella#136]
```
/Users/aeufemio/projects/ethereum/webthree-umbrella/solidity/libethereum/evmjit/libevmjit/GasMeter.cpp:169:22: error: no member named 'getNextNode' in 'llvm::Argument'
        auto cost = gasPtr->getNextNode();
                    ~~~~~~  ^
/Users/aeufemio/projects/ethereum/webthree-umbrella/solidity/libethereum/evmjit/libevmjit/Ext.cpp:67:13: error: no matching member function for call to 'SetInsertPoint'
                m_builder.SetInsertPoint(allocaIt);
                ~~~~~~~~~~^~~~~~~~~~~~~~
/usr/local/opt/llvm/include/llvm/IR/IRBuilder.h:91:8: note: candidate function not viable: no known conversion from 'llvm::ilist_iterator<llvm::Instruction>' to 'llvm::BasicBlock *' for 1st argument
  void SetInsertPoint(BasicBlock *TheBB) {
       ^
/usr/local/opt/llvm/include/llvm/IR/IRBuilder.h:98:8: note: candidate function not viable: no known conversion from 'llvm::ilist_iterator<llvm::Instruction>' to 'llvm::Instruction *' for 1st argument
  void SetInsertPoint(Instruction *I) {
       ^
/usr/local/opt/llvm/include/llvm/IR/IRBuilder.h:107:8: note: candidate function not viable: requires 2 arguments, but 1 was provided
  void SetInsertPoint(BasicBlock *TheBB, BasicBlock::iterator IP) {
       ^
/Users/aeufemio/projects/ethereum/webthree-umbrella/solidity/libethereum/evmjit/libevmjit/Array.cpp:37:25: error: no member named 'getNextNode' in 'llvm::Argument'
        auto value = arrayPtr->getNextNode();
                     ~~~~~~~~  ^
/Users/aeufemio/projects/ethereum/webthree-umbrella/solidity/libethereum/evmjit/libevmjit/Array.cpp:87:25: error: no member named 'getNextNode' in 'llvm::Argument'
        auto index = arrayPtr->getNextNode();
                     ~~~~~~~~  ^
/Users/aeufemio/projects/ethereum/webthree-umbrella/solidity/libethereum/evmjit/libevmjit/Array.cpp:111:25: error: no member named 'getNextNode' in 'llvm::Argument'
        auto index = arrayPtr->getNextNode();
                     ~~~~~~~~  ^
/Users/aeufemio/projects/ethereum/webthree-umbrella/solidity/libethereum/evmjit/libevmjit/Array.cpp:133:25: error: no member named 'getNextNode' in 'llvm::Argument'
        auto index = arrayPtr->getNextNode();
                     ~~~~~~~~  ^
/Users/aeufemio/projects/ethereum/webthree-umbrella/solidity/libethereum/evmjit/libevmjit/Array.cpp:191:27: error: no member named 'getNextNode' in 'llvm::Argument'
        auto newSize = arrayPtr->getNextNode();
                       ~~~~~~~~  ^
/Users/aeufemio/projects/ethereum/webthree-umbrella/solidity/libethereum/evmjit/libevmjit/Arith256.cpp:86:14: error: no member named 'getNextNode' in 'llvm::Argument'
        auto y = x->getNextNode();
                 ~  ^
/Users/aeufemio/projects/ethereum/webthree-umbrella/solidity/libethereum/evmjit/libevmjit/BasicBlock.cpp:170:23: error: no member named 'getNextNode' in 'llvm::Argument'
        auto sizePtr = base->getNextNode();
                       ~~~~  ^
/Users/aeufemio/projects/ethereum/webthree-umbrella/solidity/libethereum/evmjit/libevmjit/Arith256.cpp:203:17: error: no member named 'getNextNode' in 'llvm::Argument'
        auto yArg = x->getNextNode();
                    ~  ^
/Users/aeufemio/projects/ethereum/webthree-umbrella/solidity/libethereum/evmjit/libevmjit/Arith256.cpp:301:14: error: no member named 'getNextNode' in 'llvm::Argument'
        auto y = x->getNextNode();
                 ~  ^
/Users/aeufemio/projects/ethereum/webthree-umbrella/solidity/libethereum/evmjit/libevmjit/Arith256.cpp:325:14: error: no member named 'getNextNode' in 'llvm::Argument'
        auto y = x->getNextNode();
                 ~  ^
/Users/aeufemio/projects/ethereum/webthree-umbrella/solidity/libethereum/evmjit/libevmjit/Arith256.cpp:369:14: error: no member named 'getNextNode' in 'llvm::Argument'
        auto y = x->getNextNode();
                 ~  ^
/Users/aeufemio/projects/ethereum/webthree-umbrella/solidity/libethereum/evmjit/libevmjit/Arith256.cpp:415:14: error: no member named 'getNextNode' in 'llvm::Argument'
        auto y = x->getNextNode();
                 ~  ^
/Users/aeufemio/projects/ethereum/webthree-umbrella/solidity/libethereum/evmjit/libevmjit/Arith256.cpp:441:14: error: no member named 'getNextNode' in 'llvm::Argument'
        auto y = x->getNextNode();
                 ~  ^
/Users/aeufemio/projects/ethereum/webthree-umbrella/solidity/libethereum/evmjit/libevmjit/Arith256.cpp:464:25: error: no member named 'getNextNode' in 'llvm::Argument'
                auto exponent = base->getNextNode();
                                ~~~~  ^
/Users/aeufemio/projects/ethereum/webthree-umbrella/solidity/libethereum/evmjit/libevmjit/Compiler.cpp:117:4: error: no matching constructor for initialization of 'IRBuilder' (aka 'IRBuilder<>')
                        IRBuilder{it}.CreateBr(nextBlock);
                        ^        ~~~~
/usr/local/opt/llvm/include/llvm/IR/IRBuilder.h:540:7: note: candidate constructor (the implicit copy constructor) not viable: no known conversion from 'llvm::ilist_iterator<llvm::BasicBlock>' to 'const llvm::IRBuilder<true, llvm::ConstantFolder,
      llvm::IRBuilderDefaultInserter<true> >' for 1st argument
class IRBuilder : public IRBuilderBase, public Inserter {
      ^
/usr/local/opt/llvm/include/llvm/IR/IRBuilder.h:540:7: note: candidate constructor (the implicit move constructor) not viable: no known conversion from 'llvm::ilist_iterator<llvm::BasicBlock>' to
      'llvm::IRBuilder<true, llvm::ConstantFolder, llvm::IRBuilderDefaultInserter<true> >' for 1st argument
class IRBuilder : public IRBuilderBase, public Inserter {
      ^
/usr/local/opt/llvm/include/llvm/IR/IRBuilder.h:550:12: note: candidate constructor not viable: no known conversion from 'llvm::ilist_iterator<llvm::BasicBlock>' to 'llvm::LLVMContext &' for 1st argument
  explicit IRBuilder(LLVMContext &C, MDNode *FPMathTag = nullptr,
           ^
/usr/local/opt/llvm/include/llvm/IR/IRBuilder.h:560:12: note: candidate constructor not viable: no known conversion from 'llvm::ilist_iterator<llvm::BasicBlock>' to 'llvm::BasicBlock *' for 1st argument
  explicit IRBuilder(BasicBlock *TheBB, MDNode *FPMathTag = nullptr,
           ^
/usr/local/opt/llvm/include/llvm/IR/IRBuilder.h:566:12: note: candidate constructor not viable: no known conversion from 'llvm::ilist_iterator<llvm::BasicBlock>' to 'llvm::Instruction *' for 1st argument
  explicit IRBuilder(Instruction *IP, MDNode *FPMathTag = nullptr,
           ^
/usr/local/opt/llvm/include/llvm/IR/IRBuilder.h:554:12: note: candidate constructor not viable: requires at least 2 arguments, but 1 was provided
  explicit IRBuilder(BasicBlock *TheBB, const T &F, MDNode *FPMathTag = nullptr,
           ^
/usr/local/opt/llvm/include/llvm/IR/IRBuilder.h:579:3: note: candidate constructor not viable: requires at least 2 arguments, but 1 was provided
  IRBuilder(BasicBlock *TheBB, BasicBlock::iterator IP,
  ^
/usr/local/opt/llvm/include/llvm/IR/IRBuilder.h:544:3: note: candidate constructor not viable: requires at least 2 arguments, but 1 was provided
  IRBuilder(LLVMContext &C, const T &F, Inserter I = Inserter(),
  ^
/usr/local/opt/llvm/include/llvm/IR/IRBuilder.h:572:3: note: candidate constructor not viable: requires at least 3 arguments, but 1 was provided
  IRBuilder(BasicBlock *TheBB, BasicBlock::iterator IP, const T &F,
  ^
/Users/aeufemio/projects/ethereum/webthree-umbrella/solidity/libethereum/evmjit/libevmjit/Compiler.cpp:128:42: error: no viable conversion from 'llvm::ilist_iterator<llvm::BasicBlock>' to 'llvm::BasicBlock *'
                                jumpTableInput->addIncoming(destIdx, it); // Fill up PHI node
                                                                     ^~
/usr/local/opt/llvm/include/llvm/IR/Instructions.h:2569:42: note: passing argument to parameter 'BB' here
  void addIncoming(Value *V, BasicBlock *BB) {
                                         ^
1 error generated.
make[2]: *** [evmjit/libevmjit/CMakeFiles/evmjit.dir/GasMeter.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
[ 24%] Building CXX object libethcore/CMakeFiles/ethcore.dir/ChainOperationParams.cpp.o
/Users/aeufemio/projects/ethereum/webthree-umbrella/solidity/libethereum/evmjit/libevmjit/Memory.cpp:36:25: error: no member named 'getNextNode' in 'llvm::Argument'
                auto blkOffset = mem->getNextNode();
                                 ~~~  ^
/Users/aeufemio/projects/ethereum/webthree-umbrella/solidity/libethereum/evmjit/libevmjit/Memory.cpp:110:20: error: no member named 'getNextNode' in 'llvm::Argument'
        auto index = mem->getNextNode();
                     ~~~  ^
/Users/aeufemio/projects/ethereum/webthree-umbrella/solidity/libethereum/evmjit/libevmjit/Optimizer.cpp:60:35: error: no viable conversion from 'llvm::ilist_iterator<llvm::BasicBlock>' to 'llvm::BasicBlock *'
                        retPhi->addIncoming(abortCode, bbIt);
                                                       ^~~~
/usr/local/opt/llvm/include/llvm/IR/Instructions.h:2569:42: note: passing argument to parameter 'BB' here
  void addIncoming(Value *V, BasicBlock *BB) {
                                         ^
1 error generated.
make[2]: *** [evmjit/libevmjit/CMakeFiles/evmjit.dir/BasicBlock.cpp.o] Error 1
[ 24%] Building CXX object libethcore/CMakeFiles/ethcore.dir/Common.cpp.o
8 errors generated.
make[2]: *** [evmjit/libevmjit/CMakeFiles/evmjit.dir/Arith256.cpp.o] Error 1
[ 25%] Building CXX object libethcore/CMakeFiles/ethcore.dir/CommonJS.cpp.o
[ 26%] Building CXX object libethcore/CMakeFiles/ethcore.dir/ICAP.cpp.o
1 error generated.
make[2]: *** [evmjit/libevmjit/CMakeFiles/evmjit.dir/Optimizer.cpp.o] Error 1
[ 27%] Building CXX object libethcore/CMakeFiles/ethcore.dir/KeyManager.cpp.o
[ 28%] Building CXX object libethcore/CMakeFiles/ethcore.dir/Precompiled.cpp.o
1 error generated.
make[2]: *** [evmjit/libevmjit/CMakeFiles/evmjit.dir/Ext.cpp.o] Error 1
[ 29%] Building CXX object libethcore/CMakeFiles/ethcore.dir/SealEngine.cpp.o
5 errors generated.
make[2]: *** [evmjit/libevmjit/CMakeFiles/evmjit.dir/Array.cpp.o] Error 1
[ 29%] Building CXX object libethcore/CMakeFiles/ethcore.dir/Transaction.cpp.o
2 errors generated.
make[2]: *** [evmjit/libevmjit/CMakeFiles/evmjit.dir/Compiler.cpp.o] Error 1
2 errors generated.
make[2]: *** [evmjit/libevmjit/CMakeFiles/evmjit.dir/Memory.cpp.o] Error 1
make[1]: *** [evmjit/libevmjit/CMakeFiles/evmjit.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
[ 30%] Linking CXX shared library libevmcore.dylib
[ 30%] Built target evmcore
[ 31%] Linking CXX shared library libethcore.dylib
[ 31%] Built target ethcore
make: *** [all] Error 2
```

**Comment by bobsummerwill** [comment]
Yeah - I'm getting that too right now, @tymat. Sigh.

And it USED to work. I just tried the incredibly slow "brew reinstall llvm --HEAD" thing too, and still got the same result. So I think we're broken again for some reason. Please could you log a Github issue against webthree-umbrella and we'll have to triage through it collectively.

I never quite worked out the root cause of this problem, which appears NOT to be as simple as "you have too old a version of LLVM". The specific build error seems to be something to do with "auto" getting the wrong type, so the code thinks it's a list node wrapping an Argument (so you can do getNextNode() on it), but the LLVM is resolving that to the argument itself. SOMETHING is different somewhere which is resulting in different semantics. I don't know what.

LLVM is something which appears to be in dire need of turning into a sub-module, because having a dependency on "HEAD" of a moving target is even looser than the rest of the Homebrew moving-target.

**Comment by bobsummerwill** [comment]
@tymat said ...
I think llvm head on homebrew might have actually introduced this breaking change
http://llvm.org/releases/3.7.1/docs/ReleaseNotes.html not sure if this is what brew uses for HEAD

**Comment by bobsummerwill** [comment]
Or it may be some weird bootstrapping kind of issue, where we are using Clang to build LLVM and Clang?
Here's the magic - https://github.com/ethereum/homebrew-ethereum/blob/master/cpp-ethereum.rb.
Anyway - I'll just cut-and-paste this chat over to the issue. Let's carry on with discussion there, and any further participants are VERY welcome!

**Comment by bobsummerwill** [comment]
This is the magic, I meant - https://github.com/Homebrew/homebrew/blob/master/Library/Formula/llvm.rb

**Comment by bobsummerwill** [comment]
So HEAD literally just seems to mean http://llvm.org/git/llvm.git, which is mirrored at https://github.com/llvm-mirror/llvm.

ie. HEAD of master.    Totally "live", which seems to mean the prepwork for LLVM 3.8, not any specific "release".

I wonder whether, with the release of 3.7.1, that would be new enough for us now, and we could switch back to "not-HEAD"?

**Comment by bobsummerwill** [comment]
Gotta run.   Happy hunting!

**Comment by bobsummerwill** [comment]
What is your OS X version, @tymat?    And installed Clang version?

I'm 10.11.3 for the OS, and ...

```
Bobs-MacBook-Air:~ bob$ gcc --version
Configured with: --prefix=/Applications/Xcode.app/Contents/Developer/usr --with-gxx-include-dir=/usr/include/c++/4.2.1
Apple LLVM version 7.0.2 (clang-700.1.81)
Target: x86_64-apple-darwin15.3.0
Thread model: posix
Bobs-MacBook-Air:~ bob$ 
```

**Comment by bobsummerwill** [comment]
```
auto func = llvm::Function::Create(llvm::FunctionType::get(Type::Void, argTypes, false), llvm::Function::PrivateLinkage, "array.push", getModule());
func->setDoesNotThrow();
func->setDoesNotCapture(1);

auto arrayPtr = &func->getArgumentList().front();
arrayPtr->setName("arrayPtr");
auto value = arrayPtr->getNextNode();    <-- Dies here.
value->setName("value");
```

**Comment by bobsummerwill** [comment]
func should be returning Function*
func->getArgumentList() should be returning SymbolTableList<Argument>&
func->getArgumentList().front() should be returning Argument&

**Comment by bobsummerwill** [comment]
Hmm ... @chfast.    Any comment on this thread?

We get a bunch of build breaks for some still undefined set of OS X builds (including mine right now).

This getNextNode() call _does_ look suspect to me.    Not sure if LLVM changed their API at some point or if there is some other variable at play here.

But this expression really doesn't look like an intended iteration flow to me:

```
&func->getArgumentList().front()
```

**Comment by bobsummerwill** [comment]
Shouldn't it be using `func->arg_begin()` and `func->arg_end()` and `Function::arg_iterator` ?

**Comment by bobsummerwill** [comment]
Attempted fixes brewing at ...

https://github.com/doublethinkco/libethereum

... if anybody has any input?   Diffs:

https://github.com/ethereum/libethereum/compare/develop...doublethinkco:develop

**Comment by chfast** [comment]
Yes. First use LLVM 3.7 what should be `brew install llvm37`.

**Comment by bobsummerwill** [comment]
Hey @chfast!

These diffs actually fix all the build breaks and should work with old-and-new LLVM.   From what I can see, the issue is that the existing code wasn't using iterators, and got broken when the container classes changes.    So the changes below DO use iterators.    I need to review the semantics further, but these compile cleanly again against HEAD:

Now a WIP pull request - https://github.com/ethereum/libethereum/pull/157.

I am just looking at `llvm37` now, and seeing if I can do a build against llvm37 too with them, which I think should be possible.     The issue I hit first with `brew install llvm37` though, is that bottle for El Capitan doesn't contain the cmake files in the share folder (and looks a little incomplete in general).    So I am now doing a `brew install -s llvm37`.     I have removed the `--with-clang` shown in the OS X build instructions, because I am not convinced that is needed.    It will break in an obvious way if it is, and I'll re-add it.

The other thing I wanted to ask you is how you had built using `llvm37`, because when I install it, it goes into a co-directory of llvm, so the CMake steps don't find it for the main webthree-umbrella build.   Perhaps you are building in another way, or are building `evmjit` repo on its own with a direct CMake?

Just wondering your thoughts on how we would plug `llvm37` into the build system, if we wanted to switch to that stable version (which I think is a good idea, especially now there has been a 3.7.1 release which is now used in the homebrew recipe).

**Comment by bobsummerwill** [comment]
`brew install llvm37 --with-clang` doesn't create that cmake share folder.   Trying `brew install -s llvm37 --with-clang` now.

**Comment by chfast** [comment]
Using iterators is good idea. Just try convert them to pointers straight away or use plain interators. LLVM iterators had weird property that inter<T> is implicitly convertible to T*.

**Comment by bobsummerwill** [comment]
Thanks, @chfast.    I guess that the removed that implicit conversion between LLVM 3.7.1 and HEAD, and that was the source of the break.    And use of `auto` for this code made it very prone to breaks when LLVM itself changed semantics.

`brew install -s llvm37 --with-clang` didn't have the CMake share folder either, so I don't think it's worth bothering trying to lock to 3.7.1.

**Comment by bobsummerwill** [comment]
@tymat Want to give https://github.com/ethereum/libethereum/pull/157 a try, and verify it fixes the build issues on your machine too?

**Comment by chfast** [comment]
I think using HEAD (i.e. 3.8+) would be a mistake. I had to do a lot of fixes for 3.7 to make it ready for Ethereum. And we don't have nothing more than regression tests checked currently. If homebrew fucked up LLVM again, ask them kindly to fix it.

**Comment by bobsummerwill** [comment]
The build instructions for OS X have **always** been using HEAD.   Not sure when those were originally written.   I think I first used those back in Sep/Oct 2015, at least.

We should look into what the buildbot was doing when that was active, and what the Jenkins is doing now.

Perhaps your setup is atypical, @chfast?    When did you last do a fresh build setup?    Or have you just had the same Mac setup for months?    Did you do a `brew install llvm37` yourself at some point in the past?   What does your `/local/opt/llvm37` folder look like?
